idea/src/org/jetbrains/kotlin/idea/quickfix/expectactual/ExpectActualUtils.kt

```diff
@@ -47,26 +75,252 @@ fun createFileForDeclaration(module: Module, declaration: KtNamedDeclaration): K
+    generatedClass: KtClassOrObject? = null,
+    outerClasses: List<KtClassOrObject> = emptyList()
+): KtFunction {
+    if (generateExpect) {
+        val accessibleClasses = outerClasses + listOfNotNull(generatedClass)
+        descriptor.checkTypeParameterBoundsAccessibility(accessibleClasses)
+        descriptor.returnType?.checkAccessibility(accessibleClasses)
+        descriptor.valueParameters.forEach {
+            it.type.checkAccessibility(accessibleClasses)
+        }
+    }
+    val memberChooserObject = create(
+        originalFunction, descriptor, descriptor,
+        if (generateExpect || descriptor.modality == Modality.ABSTRACT) NO_BODY else EMPTY_OR_TEMPLATE
+    )
+    return memberChooserObject.generateMember(
+        targetClass = generatedClass,
+        copyDoc = true,
+        project = project,
+        forceActual = !generateExpect,
+        forceExpect = generateExpect
+    ) as KtFunction
+}
+
+internal fun generateProperty(
+    project: Project,
+    generateExpect: Boolean,
+    originalProperty: KtNamedDeclaration,
+    descriptor: PropertyDescriptor,
+    generatedClass: KtClassOrObject? = null,
+    outerClasses: List<KtClassOrObject> = emptyList()
+): KtProperty {
+    if (generateExpect) {
+        val accessibleClasses = outerClasses + listOfNotNull(generatedClass)
+        descriptor.checkTypeParameterBoundsAccessibility(accessibleClasses)
+        descriptor.type.checkAccessibility(accessibleClasses)
+    }
+    val memberChooserObject = create(
+        originalProperty, descriptor, descriptor,
+        if (generateExpect || descriptor.modality == Modality.ABSTRACT) NO_BODY else EMPTY_OR_TEMPLATE
+    )
+    return memberChooserObject.generateMember(
+        targetClass = generatedClass,
+        copyDoc = true,
+        project = project,
+        forceActual = !generateExpect,
+        forceExpect = generateExpect
+    ) as KtProperty
+}
+
+private fun CallableMemberDescriptor.checkTypeParameterBoundsAccessibility(accessibleClasses: List<KtClassOrObject>) {
+    typeParameters.forEach { typeParameter ->
+        typeParameter.upperBounds.forEach { upperBound ->
+            upperBound.checkAccessibility(accessibleClasses)
+        }
+    }
+}
+
+private fun KotlinType.checkAccessibility(accessibleClasses: List<KtClassOrObject>) {
+    for (argument in arguments) {
+        if (argument.isStarProjection) continue
+        argument.type.checkAccessibility(accessibleClasses)
+    }
+    val classifierDescriptor = constructor.declarationDescriptor as? ClassifierDescriptorWithTypeParameters ?: return
+    val moduleDescriptor = classifierDescriptor.module
+    if (moduleDescriptor.getMultiTargetPlatform() == MultiTargetPlatform.Common) {
+        // Common classes are Ok; unfortunately this check does not work correctly for simple (non-expect) classes from common module
+        return
+    }
+    val declaration = DescriptorToSourceUtils.descriptorToDeclaration(classifierDescriptor)
```

Comment:
Shouldn't we use `DescriptorToSourceUtilsIde` here?

Reply:
I just don't exactly know the difference :). Mostly I use just `DescriptorToSourceUtils`.